Add Ruby support

calebmeyer opened the pull request:
#### Description

Add a new prompt section (Tide Item) for Ruby. This should support package manager installs and Rbenv/RVM, though I've only tested with RVM

#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/3342777/147694156-ee47d63a-54bf-4fec-8d30-e05ebfd119f0.png)

#### How Has This Been Tested

- [ ] I have tested using **Linux**.
- [x] I have tested using **MacOS**.

![image](https://user-images.githubusercontent.com/3342777/147694420-db9bb345-9db4-4fd1-97e1-fdc16b157ce0.png)


#### Checklist

- [ ] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

calebmeyer commented:
I didn't check the documentation box because the docs are a wiki and I figured it would be better to update after the PR is merged.

calebmeyer commented:
It appears the lint failure is unrelated to these changes.

calebmeyer commented on functions/tide/configure/configs/classic.fish:
```diff
@@ -75,6 +75,9 @@ tide_right_prompt_suffix ''
 tide_rustc_color F74C00
 tide_rustc_icon ''
+tide_rvm_bg_color 444444
+tide_rvm_color CC342D
```
This is a dark red which I took from the ruby language website

IlanCosman commented:
Thanks for the PR Caleb 😄 

1. I think we should do what we did for `node` here. I.E. instead of an item for each version manager, have a `ruby` item that prints `ruby --version`. Assuming your version manager is set up correctly, this should do the same thing. You can code that up if you wish, or I can 👍

2. Thanks for adding that Wiki page. I was going to do something similar eventually, I swear 😅

3. Correcting a small misconception: `_tide_print_item` has no knowledge about icons. Instead we take advantage of [Fish's cartesian expansion system](https://fishshell.com/docs/current/language.html?highlight=cartesian#combining-lists-cartesian-product), which makes it so that `printf %s $var_that_is_empty_or_unset"hello"` prints nothing. So `_tide_print_item` doesn't ever ever consider icons, it just sees text or nothing.

calebmeyer commented:
> `_tide_print_item` has no knowledge about icons

That definitely was a misconception for me. I wondered about the syntax. It almost looked like a mistake to have `$icon' '`, but it worked, and it stopped working if I added a space between them so I assumed it was something print item was doing for me.

I've changed it over to ruby (https://github.com/calebmeyer/tide/commit/c7b674b63935e9c497e9a56699512b2fac3d8634), but locally all my tests (`make`) are failing with some indecipherable unicode stuff (apparently we're getting an extra character somehow?):
![image](https://user-images.githubusercontent.com/3342777/147953458-25c924fe-e914-44ac-9c87-67c7b21c3b13.png)

calebmeyer commented:
Okay, looks like the tests are happy again. Not sure why they stopped working for me locally. I may have to uninstall/reinstall to be able to use them.

pusewicz commented:
I believe this should now also be removed? https://github.com/IlanCosman/tide/blob/main/functions/_tide_item_chruby.fish

halostatue commented on functions/_tide_item_ruby.fish:
```diff
@@ -0,0 +1,5 @@
+function _tide_item_ruby
+    if test -e .ruby-version
```
As someone who would use this, a better choice would be to `test -e Gemfile || test -e Rakefile || test -e .ruby-version`. Not all of my Ruby projects have a `.ruby-version` file. Most of my projects have a Gemfile or Rakefile (you could also check for a `.gemspec`).

smook1980 replied:
I'd also like to advocate for testing for `Gemfile`, `Rakefile` or *.gemspec` along with `ruby --version` as it would support projects without a version specified as well as ASDF's `.tool-versions` file. Also, thanks for doing the work to add ruby support!

cmer replied:
I agree, I believe this line should be:

`if test -e Gemfile || test -e Rakefile || test -e *.gemspec || test -e .ruby-version || test -e .tool-versions`

Would love to see the code updated for for this feature to be merged!

IlanCosman commented:
Yah sorry for not moving on this. It's been my intention to wait for 6.0 on this because it is a breaking change in that it will involve removing `chruby`.

rrevi commented:
Thanks @calebmeyer for this MR! 

You motivated me to create a [PR](https://github.com/IlanCosman/tide/pull/322) to add Crystal support.

However, I am having a little of bit of trouble testing live. I was trying to re-create the screenshots you have here, but after running `make`, I exit my Fish shell, get back in to Fish, move to a directory with a Crystal project file and my tide prompt for Crystal doesn't show. This is all despite that the `make test` for Crystal passes. Maybe you can help me test?

rrevi commented:
does it make sense to change `chruby` in the [_tide_remove_unusable_items.fish](https://github.com/calebmeyer/tide/blob/add-rvm/functions/_tide_remove_unusable_items.fish) file?

cmer commented:
Could we get this merged, @IlanCosman ?

cmer commented:
I have created a new version of this including test at #437. Inspired by the work in this PR. Would love feedback.

IlanCosman commented:
@cmer Thanks for the PR, I'll take a look at it ~1 week from now when I have time :)

IlanCosman commented:
Closing this in favor of #437
